.github/workflows/build.yml

```diff
@@ -25,60 +25,116 @@ jobs:
+          . venv-gusto/bin/activate
+          pip install ./gusto-repo
+          pip list
+
+      - name: Test serial-only netCDF
         run: |
-          . /home/firedrake/firedrake/bin/activate
-          firedrake-clean
-          export GUSTO_PARALLEL_LOG=FILE
-          export PYOP2_CFLAGS=-O0
-          python -m pip uninstall -y netCDF4
-          export HDF5_DIR=$PETSC_DIR/packages
-          export NETCDF4_DIR=$PETSC_DIR/packages
-          python -m pip install --no-binary netCDF4 --no-build-isolation netCDF4
-          python -m pytest \
-            -n 12 --dist worksteal \
-            --durations=100 \
-            --timeout=3600 \
-            --timeout-method=thread \
-            -o faulthandler_timeout=3660 \
-            -v unit-tests integration-tests examples
-        timeout-minutes: 120
+          . venv-gusto/bin/activate
+          : # Run the serial tests
+          firedrake-run-split-tests 1 1 -n 12 --verbose gusto-repo/integration-tests/model/test_nc_outputting.py
+          : # Run the parallel tests
+          firedrake-run-split-tests 2 6 --verbose gusto-repo/integration-tests/model/test_nc_outputting.py
+        timeout-minutes: 10
+
+      - name: Install parallel netCDF
+        # Run even if the step above failed
+        if: success() || steps.install-one.conclusion == 'success'
+        id: install-two
+        run: |
+          . venv-gusto/bin/activate
+          pip uninstall -y netCDF4
+          : # '--no-build-isolation' has to be passed for mpi4py to be found
+          : # but this means that we need to install any extra build
+          : # dependencies manually
+          pip install Cython
+          pip install --no-binary netCDF4 --no-build-isolation netCDF4
+
+      - name: Run tests (nprocs = 1)
+        run: |
+          . venv-gusto/bin/activate
+          : # Use pytest-xdist here so we can have a single collated output (not possible
+          : # for parallel tests)
+          firedrake-run-split-tests 1 1 -n 12 --dist worksteal "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 60
+
+      - name: Run tests (nprocs = 2)
+        # Run even if earlier tests failed
+        if: success() || steps.install-two.conclusion == 'success'
+        run: |
+          . venv-gusto/bin/activate
+          firedrake-run-split-tests 2 6 "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 30
+
+      - name: Run tests (nprocs = 3)
+        if: success() || steps.install-two.conclusion == 'success'
+        run: |
+          . venv-gusto/bin/activate
+          firedrake-run-split-tests 3 4 "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 10
+
+      - name: Run tests (nprocs = 4)
+        if: success() || steps.install-two.conclusion == 'success'
+        run: |
+          . venv-gusto/bin/activate
+          firedrake-run-split-tests 4 3 "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 10
+
       - name: Prepare logs
-        if: always()
+        if: success() || steps.install-two.conclusion == 'success'
```

Comment:
I don't think this step should be changed from `always()`, if an above step fails it doesn't matter if this step fails (likewise below for the upload)

Reply:
The logic here is saying to only try and upload the pytest logs if we successfully install Gusto. That seems like the right thing to do?

Reply:
You are running tests before "install 2" though

Reply:
It's more complicated now that we call pytest lots of different times. I guess we could upload the logs for the initial tests but there are so few tests being run and the same information is available in stdout. The logs are mostly helpful for when stdout is too polluted to understand.

Reply:
Gusto logs contain information that is not printed to stdout and is useful for debugging

Reply:
And were those previously found in `/tmp/pytest-of-firedrake/...`? I am confused about whether these log files we are uploading are the output from pytest or the output of something else.

Reply:
Resolving as this is the same discussion as another thread.